File: app/src/main/java/zac4j/com/webapp/WebJsBindActivity.java
package zac4j.com.webapp;

import android.annotation.SuppressLint;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.webkit.JavascriptInterface;
import android.webkit.JsResult;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.widget.Button;
import android.widget.EditText;

/**
 * Activity for JavaScript Binding
 * Created by Administrator on 2017/7/21.
 */

public class WebJsBindActivity extends AppCompatActivity {

  private JavascriptCaller mCaller = new JavascriptCaller();

  private EditText mNameField;
  private WebView mBrowser;

  @SuppressLint("SetJavaScriptEnabled") @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_js_bind);

    mNameField = (EditText) findViewById(R.id.jsbind_et_name_field);
    findViewById(R.id.jsbind_btn_jscaller).setOnClickListener(new View.OnClickListener() {
      @Override public void onClick(View v) {
        callJsHelloMethod(getUsername(mNameField));
      }
    });
    mBrowser = (WebView) findViewById(R.id.webview);

    // enable javascript
    mBrowser.getSettings().setJavaScriptEnabled(true);
    // enable zoom control
    mBrowser.getSettings().setBuiltInZoomControls(true);
    mBrowser.setWebChromeClient(new MyWebChromeClient());
    mBrowser.addJavascriptInterface(mCaller, "jscaller");
    mBrowser.loadUrl("file:///android_asset/jsbind.html");
  }

  private void callJsHelloMethod(String username) {
    mBrowser.loadUrl("javascript:sayHelloFromJS('" + username + "')");
  }

  private String getUsername(EditText editText) {
    return editText == null ? "" : mNameField.getText().toString();
  }

  private class JavascriptCaller {
    @JavascriptInterface public String getNameFromEditText() {
      return mNameField.getText().toString();
    }

    @JavascriptInterface public void sayHelloFromAndroid(String actualData) {
      new AlertDialog.Builder(WebJsBindActivity.this).setTitle(
          "Android method called from JavaScript!")
          .setMessage("Android says:Hello " + actualData + " !!! Are you OK?")
          .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
            @Override public void onClick(DialogInterface dialog, int which) {
              //
            }
          })
          .setCancelable(false)
          .create()
          .show();
    }
  }

  private class MyWebChromeClient extends WebChromeClient {
    @Override
    public boolean onJsAlert(WebView view, String url, String message, final JsResult result) {
      new AlertDialog.Builder(WebJsBindActivity.this).setTitle("WebApp Alert")
          .setMessage(message)
          .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
            @Override public void onClick(DialogInterface dialog, int which) {
              result.confirm();
            }
          })
          .setCancelable(false)
          .create()
          .show();
      return true;
    }
  }
}

File: app/src/main/java/zac4j/com/webapp/WebJsBindActivity.kt
package zac4j.com.webapp

import android.R.string
import android.annotation.SuppressLint
import android.content.DialogInterface
import android.content.DialogInterface.OnClickListener
import android.os.Bundle
import android.util.Log
import android.view.View
import android.webkit.JavascriptInterface
import android.webkit.JsResult
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.EditText
import androidx.appcompat.app.AlertDialog.Builder
import androidx.appcompat.app.AppCompatActivity
import zac4j.com.webapp.R.id
import zac4j.com.webapp.R.layout

/**
 * Activity for JavaScript Binding
 * Created by Administrator on 2017/7/21.
 */
class WebJsBindActivity() : AppCompatActivity() {
  private val mCaller = JavascriptCaller()

  /**
   * NameInput enter your name
   */
  private lateinit var mNameInput: EditText

  /**
   * Web Container
   */
  private lateinit var mBrowser: WebView

  @SuppressLint("SetJavaScriptEnabled")
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(layout.activity_js_bind)
    mNameInput = findViewById<View>(id.jsbind_et_name_field) as EditText
    findViewById<View>(
        id.jsbind_btn_jscaller
    ).setOnClickListener(
        View.OnClickListener { callJsHelloMethod(getUsername(mNameInput)) })
    mBrowser = findViewById<View>(id.webview) as WebView

    // enable javascript
    mBrowser.settings.javaScriptEnabled = true
    // enable zoom control
    mBrowser.settings.builtInZoomControls = true
    mBrowser.settings.allowUniversalAccessFromFileURLs
    mBrowser.webChromeClient = MyWebChromeClient()
    // add custom js interface
    mBrowser.addJavascriptInterface(mCaller, "jsCaller")
    mBrowser.loadUrl("file:///android_asset/jsbind.html")
  }

  private fun callJsHelloMethod(username: String) {
    //mBrowser.loadUrl("javascript:sayHelloFromJS('$username')")
    mBrowser.evaluateJavascript("sayHelloFromJS('$username')") {
      Log.i("ZacLog", "callJsHelloMethod: ")
    }
  }

  private fun getUsername(editText: EditText?): String {
    return if (editText == null) "" else mNameInput.text
        .toString()
  }

  private inner class JavascriptCaller() {
    @get:JavascriptInterface val nameFromEditText: String
      get() = mNameInput.text
          .toString()

    @JavascriptInterface fun sayHelloFromAndroid(actualData: String) {
      Builder(this@WebJsBindActivity)
          .setTitle(
              "Android method called from JavaScript!"
          )
          .setMessage("Android says:Hello $actualData !!! This call from Browser")
          .setPositiveButton(string.ok) { _, _ ->
            //
          }
          .setCancelable(false)
          .create()
          .show()
    }
  }

  private inner class MyWebViewClient() : WebViewClient() {
    override fun shouldOverrideUrlLoading(
      view: WebView?,
      request: WebResourceRequest?
    ): Boolean {
      return super.shouldOverrideUrlLoading(view, request)
    }
  }

  private inner class MyWebChromeClient() : WebChromeClient() {
    override fun onJsAlert(
      view: WebView,
      url: String,
      message: String,
      result: JsResult
    ): Boolean {
      Builder(this@WebJsBindActivity)
          .setTitle("WebApp Alert")
          .setMessage(message)
          .setPositiveButton(
              string.ok
          ) { _, _ -> result.confirm() }
          .setCancelable(false)
          .create()
          .show()
      return true
    }
  }
}